### ISSUE TITLE
Issue with the performance of updating tensors on TPUs

### ISSUE THREAD

**Issue by ryanwongsa** [pytorch/xla#2365]
## ❓ Questions and Help

Example code:

```python
import torch
import torch_xla
import torch_xla.core.xla_model as xm

x = torch.randn((128, 1, 300), device=xm.xla_device())
y = torch.randn((128, 1, 300), device=xm.xla_device())

x[x>0.0] = y[x>0.0]  # slow here
```

The last line seems to have significant performance issues compared to running on a CPU and GPU. Is there are reason for the slow performance or alternative method to update parts of tensors while on a TPU?

**Comment by JackCaoG** [comment]
I run you code and print the metric
```
Metric: CompileTime
  TotalSamples: 1
  Accumulator: 083ms723.862us
  Percentiles: 1%=083ms723.862us; 5%=083ms723.862us; 10%=083ms723.862us; 20%=083ms723.862us; 50%=083ms723.862us; 80%=083ms723.862us; 90%=083ms723.862
us; 95%=083ms723.862us; 99%=083ms723.862us
Metric: DeviceLockWait
  TotalSamples: 2
  Accumulator: 034.136us
  ValueRate: 271.993us / second
  Rate: 15.9358 / second
  Percentiles: 1%=014.036us; 5%=014.036us; 10%=014.036us; 20%=014.036us; 50%=020.100us; 80%=020.100us; 90%=020.100us; 95%=020.100us; 99%=020.100us
Metric: ExecuteTime
  TotalSamples: 2
  Accumulator: 012ms495.414us
  ValueRate: 372ms609.212us / second
  Rate: 59.4793 / second
  Percentiles: 1%=002ms220.637us; 5%=002ms220.637us; 10%=002ms220.637us; 20%=002ms220.637us; 50%=010ms274.777us; 80%=010ms274.777us; 90%=010ms274.777
us; 95%=010ms274.777us; 99%=010ms274.777us
Metric: InboundData
  TotalSamples: 2
  Accumulator: 75.00KB
  ValueRate: 3.42MB / second
  Rate: 93.428 / second
  Percentiles: 1%=37.50KB; 5%=37.50KB; 10%=37.50KB; 20%=37.50KB; 50%=37.50KB; 80%=37.50KB; 90%=37.50KB; 95%=37.50KB; 99%=37.50KB
Metric: InputOutputAliasCount
  TotalSamples: 1
  Accumulator: 1.00
  Percentiles: 1%=1.00; 5%=1.00; 10%=1.00; 20%=1.00; 50%=1.00; 80%=1.00; 90%=1.00; 95%=1.00; 99%=1.00
Metric: IrValueTensorToXlaData
  TotalSamples: 2
  Accumulator: 010ms250.578us
  ValueRate: 584ms189.997us / second
  Rate: 113.982 / second
  Percentiles: 1%=003ms415.493us; 5%=003ms415.493us; 10%=003ms415.493us; 20%=003ms415.493us; 50%=007ms835.085us; 80%=007ms835.085us; 90%=007ms835.085
us; 95%=007ms835.085us; 99%=007ms835.085us
Metric: OutboundData
  TotalSamples: 5
  Accumulator: 901.10KB
  ValueRate: 6.25MB / second
  Rate: 35.4877 / second
  Percentiles: 1%=8.00B; 5%=8.00B; 10%=8.00B; 20%=8.00B; 50%=8.00B; 80%=450.54KB; 90%=450.54KB; 95%=450.54KB; 99%=450.54KB
Metric: ReleaseDataHandlesTime
  TotalSamples: 2
  Accumulator: 008ms328.663us
  ValueRate: 575ms603.155us / second
  Rate: 137.982 / second
  Percentiles: 1%=786.118us; 5%=786.118us; 10%=786.118us; 20%=786.118us; 50%=008ms542.545us; 80%=008ms542.545us; 90%=008ms542.545us; 95%=008ms542.545
us; 99%=008ms542.545us
Metric: TensorsGraphSize
  TotalSamples: 2
  Accumulator: 24.00
  ValueRate: 575.42 / second
  Rate: 47.9519 / second
  Percentiles: 1%=12.00; 5%=12.00; 10%=12.00; 20%=12.00; 50%=12.00; 80%=12.00; 90%=12.00; 95%=12.00; 99%=12.00
Metric: TransferFromServerTime
  TotalSamples: 2
  Accumulator: 021ms630.577us
  ValueRate: 964ms694.986us / second
  Rate: 93.4239 / second
  Percentiles: 1%=004ms260.155us; 5%=004ms260.155us; 10%=004ms260.155us; 20%=004ms260.155us; 50%=016ms370.422us; 80%=016ms370.422us; 90%=016ms370.422
us; 95%=016ms370.422us; 99%=016ms370.422us
Metric: TransferToServerTime
  TotalSamples: 5
  Accumulator: 021ms640.978us
  ValueRate: 149ms263.466us / second
  Rate: 36.1571 / second
```
You can see that `CompileTime` contribute to the most of the slowness, and `ExecuteTime` is actually only `012ms`. This is due to our relatively complex lowering of this index input with condition. You can use 

```
print(torch_xla._XLAC._get_xla_tensors_text([x]))
torch_xla._XLAC._get_xla_tensors_hlo(x)
```

to view the IR graph and hlo. The good news is that you only need to compile once(if input shape does not change), so this slowness is most likely a one time thing.

**Comment by ryanwongsa** [comment]
I see, thanks. The problem is that my use case requires this computation to be done quite often with different input shapes and new tensors each time. In my above example the shape is (128,1,300) but the first part might change to for example (100,1,300). Does that mean my computation will always be slow?

**Comment by JackCaoG** [comment]
In genera if you want to achieve good performance on the pytorch/xla you want to stabilize your graph within a few steps. More details on [here](https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats).

**Comment by ryanwongsa** [comment]
Alright, thanks for the link and the help.
